File: libs/document-processor-next/src/runtime/document-processing-runtime.ts
import { Blue, BlueNode } from '@blue-labs/language';
import type { JsonPatch } from '../model/shared/json-patch.js';
import { PatchEngine, type PatchResult } from './patch-engine.js';
import { EmissionRegistry } from './emission-registry.js';
import { GasMeter } from './gas-meter.js';
import { ScopeRuntimeContext } from './scope-runtime-context.js';

export type DocumentUpdateData = PatchResult;

export class DocumentProcessingRuntime {
  private readonly patchEngine: PatchEngine;
  private readonly emissionRegistry = new EmissionRegistry();
  private readonly meter: GasMeter;
  private runTerminated = false;

  constructor(
    private readonly documentRef: BlueNode,
    private readonly blueRef: Blue,
  ) {
    this.patchEngine = new PatchEngine(this.documentRef);
    this.meter = new GasMeter(this.blueRef);
  }

  document(): BlueNode {
    return this.documentRef;
  }

  scopes(): Map<string, ScopeRuntimeContext> {
    return this.emissionRegistry.scopes();
  }

  scope(scopePath: string): ScopeRuntimeContext {
    return this.emissionRegistry.scope(scopePath);
  }

  existingScope(scopePath: string): ScopeRuntimeContext | undefined {
    return this.emissionRegistry.existingScope(scopePath);
  }

  rootEmissions(): readonly BlueNode[] {
    return this.emissionRegistry.rootEmissions();
  }

  recordRootEmission(emission: BlueNode): void {
    this.emissionRegistry.recordRootEmission(emission);
  }

  addGas(amount: number): void {
    this.meter.add(amount);
  }

  blue(): Blue {
    return this.blueRef;
  }

  gasMeter(): GasMeter {
    return this.meter;
  }

  totalGas(): number {
    return this.meter.totalGas();
  }

  isRunTerminated(): boolean {
    return this.runTerminated;
  }

  markRunTerminated(): void {
    this.runTerminated = true;
  }

  isScopeTerminated(scopePath: string): boolean {
    return this.emissionRegistry.isScopeTerminated(scopePath);
  }

  directWrite(path: string, value: BlueNode | null): void {
    this.patchEngine.directWrite(path, value);
  }

  applyPatch(originScopePath: string, patch: JsonPatch): DocumentUpdateData {
    return this.patchEngine.applyPatch(originScopePath, patch);
  }
}

File: references/java-sdk/src/main/java/blue/language/processor/DocumentProcessingRuntime.java
package blue.language.processor;

import blue.language.model.Node;
import blue.language.processor.model.JsonPatch;
import java.math.BigInteger;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * Runtime state holder for a single document-processing invocation.
 */
public final class DocumentProcessingRuntime {

    private final Node document;
    private final PatchEngine patchEngine;
    private final EmissionRegistry emissionRegistry;
    private final GasMeter gasMeter;
    private boolean runTerminated;

    public DocumentProcessingRuntime(Node document) {
        this.document = Objects.requireNonNull(document, "document");
        this.patchEngine = new PatchEngine(this.document);
        this.emissionRegistry = new EmissionRegistry();
        this.gasMeter = new GasMeter();
    }

    public Node document() {
        return document;
    }

    public Map<String, ScopeRuntimeContext> scopes() {
        return Collections.unmodifiableMap(emissionRegistry.scopes());
    }

    public ScopeRuntimeContext scope(String scopePath) {
        return emissionRegistry.scope(scopePath);
    }

    public ScopeRuntimeContext existingScope(String scopePath) {
        return emissionRegistry.existingScope(scopePath);
    }

    public List<Node> rootEmissions() {
        return emissionRegistry.rootEmissions();
    }

    public void recordRootEmission(Node emission) {
        emissionRegistry.recordRootEmission(emission);
    }

    public void addGas(long amount) {
        gasMeter.add(amount);
    }

    public long totalGas() {
        return gasMeter.totalGas();
    }

    public void chargeScopeEntry(String scopePath) {
        gasMeter.chargeScopeEntry(scopePath);
    }

    public void chargeInitialization() {
        gasMeter.chargeInitialization();
    }

    public void chargeChannelMatchAttempt() {
        gasMeter.chargeChannelMatchAttempt();
    }

    public void chargeHandlerOverhead() {
        gasMeter.chargeHandlerOverhead();
    }

    public void chargeBoundaryCheck() {
        gasMeter.chargeBoundaryCheck();
    }

    public void chargePatchAddOrReplace(Node value) {
        gasMeter.chargePatchAddOrReplace(value);
    }

    public void chargePatchRemove() {
        gasMeter.chargePatchRemove();
    }

    public void chargeCascadeRouting(int scopeCount) {
        gasMeter.chargeCascadeRouting(scopeCount);
    }

    public void chargeEmitEvent(Node event) {
        gasMeter.chargeEmitEvent(event);
    }

    public void chargeBridge(Node event) {
        gasMeter.chargeBridge(event);
    }

    public void chargeDrainEvent() {
        gasMeter.chargeDrainEvent();
    }

    public void chargeCheckpointUpdate() {
        gasMeter.chargeCheckpointUpdate();
    }

    public void chargeTerminationMarker() {
        gasMeter.chargeTerminationMarker();
    }

    public void chargeLifecycleDelivery() {
        gasMeter.chargeLifecycleDelivery();
    }

    public void chargeFatalTerminationOverhead() {
        gasMeter.chargeFatalTerminationOverhead();
    }

    public void chargeTriggerEventBase() {
        gasMeter.chargeTriggerEventBase();
    }

    public void chargeUpdateDocumentBase(int changesetLength) {
        gasMeter.chargeUpdateDocumentBase(changesetLength);
    }

    public void chargeDocumentSnapshot(String absolutePointer, Node snapshot) {
        gasMeter.chargeDocumentSnapshot(absolutePointer, snapshot);
    }

    public void chargeWasmGas(BigInteger wasmFuel) {
        gasMeter.chargeWasmGas(wasmFuel);
    }

    public void chargeWasmGas(long wasmFuel) {
        gasMeter.chargeWasmGas(wasmFuel);
    }

    public boolean isRunTerminated() {
        return runTerminated;
    }

    public void markRunTerminated() {
        runTerminated = true;
    }

    public boolean isScopeTerminated(String scopePath) {
        return emissionRegistry.isScopeTerminated(scopePath);
    }

    public void directWrite(String path, Node value) {
        patchEngine.directWrite(path, value);
    }

    public DocumentUpdateData applyPatch(String originScopePath, JsonPatch patch) {
        PatchEngine.PatchResult result = patchEngine.applyPatch(originScopePath, patch);
        return new DocumentUpdateData(result.path(),
                result.before(),
                result.after(),
                result.op(),
                result.originScope(),
                result.cascadeScopes());
    }

    static final class DocumentUpdateData {
        private final String path;
        private final Node before;
        private final Node after;
        private final JsonPatch.Op op;
        private final String originScope;
        private final List<String> cascadeScopes;

        DocumentUpdateData(String path,
                           Node before,
                           Node after,
                           JsonPatch.Op op,
                           String originScope,
                           List<String> cascadeScopes) {
            this.path = path;
            this.before = before;
            this.after = after;
            this.op = op;
            this.originScope = originScope;
            this.cascadeScopes = cascadeScopes;
        }

        String path() {
            return path;
        }

        Node before() {
            return before;
        }

        Node after() {
            return after;
        }

        JsonPatch.Op op() {
            return op;
        }

        String originScope() {
            return originScope;
        }

        List<String> cascadeScopes() {
            return cascadeScopes;
        }
    }
}
